File: acme-admin/src/main/java/com/acme/admin/controller/BudgetController.java
package com.acme.admin.controller;

import com.acme.admin.domain.BudgetEntity;
import com.acme.admin.dto.BudgetCreateRequest;
import com.acme.admin.service.BudgetService;
import jakarta.validation.Valid;
import java.util.List;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/api/finance/budgets")
public class BudgetController {
  private final BudgetService service;

  public BudgetController(BudgetService service) {
    this.service = service;
  }

  @GetMapping
  public List<BudgetEntity> list() {
    return service.list();
  }

  @PostMapping
  public BudgetEntity create(@Valid @RequestBody BudgetCreateRequest req) {
    return service.create(req);
  }
}

File: acme-admin-dotnet/Controllers/BudgetController.cs
using Acme.Admin.Api.Domain;
using Acme.Admin.Api.DTO;
using Acme.Admin.Api.Security;
using Acme.Admin.Api.Services;
using Acme.Admin.Api.Validation;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace Acme.Admin.Api.Controllers;

[ApiController]
[Route("api/finance/budgets")]
public sealed class BudgetController(BudgetService service) : ControllerBase
{
    [HttpGet]
    [Authorize(Policy = PermissionPolicies.FinanceRead)]
    public async Task<ActionResult<PagedResult<BudgetEntity>>> List(
        [FromQuery] int page = Paging.DefaultPage,
        [FromQuery] int size = Paging.DefaultSize,
        CancellationToken ct = default)
    {
        var (p, s) = Paging.Normalize(page, size);
        return Ok(await service.ListAsync(p, s, ct));
    }

    [HttpPost]
    [Authorize(Policy = PermissionPolicies.FinanceWrite)]
    public async Task<ActionResult<BudgetEntity>> Create([FromBody] BudgetCreateRequest req, CancellationToken ct)
    {
        return Ok(await service.CreateAsync(req, ct));
    }
}
